### PR TITLE
ExecuteToEither

### PR DESCRIPTION
**Pull request by bonahona**
Added new worker functions, combining map() and executeToResult()

### CODE REVIEW

File: src/main/kotlin/JobState.kt [hunk 1]
```diff
@@ -19,6 +19,19 @@ class JobStateException(
     innerException: Throwable
 ):Exception(message, innerException)
 
+class Either<out L> (
```
Comment (williamhogman) [inline src/main/kotlin/JobState.kt]: Needs to switch name as Either<L, R> is another thing

File: src/main/kotlin/JobState.kt [hunk 2]
```diff
@@ -19,6 +19,19 @@ class JobStateException(
...
+    val left: L? = null,
+    val right: WorkerResult
+) {
+    fun isLeft() = left != null
+
+    companion object {
```
Comment (williamhogman) [inline src/main/kotlin/JobState.kt]: Not really a either if both can be non-null

File: src/main/kotlin/JobState.kt (hunk 1 above)
Comment (williamhogman) [reply src/main/kotlin/JobState.kt]: Or just make it a real Either...
Usually this data-type is called a result.
Eitherway it seems sane.

File: src/main/kotlin/JobState.kt [hunk 3]
```diff
@@ -189,6 +205,34 @@ class JobState<U: Any> constructor(val value: U?, val interceptors: List<WorkerI
         return process(lastInterceptor, JobStatus.TransientFailure)
     }
 
+    private suspend fun <R: Any> processEither(fn: suspend(U) -> Either<R>, msg: String? = null): JobState<R> {
```
Comment (williamhogman) [inline src/main/kotlin/JobState.kt]: Function last?

File: src/main/kotlin/JobState.kt [hunk 4]
```diff
@@ -53,6 +66,10 @@ class JobState<U: Any> constructor(val value: U?, val interceptors: List<WorkerI
     suspend fun executeToResult(fn: suspend (U) -> WorkerResult) = processWorkerFunction(fn)
     suspend fun executeToResult(msg: String, fn: suspend (U) -> WorkerResult) = processWorkerFunction(fn, msg)
 
+
+    suspend fun <R: Any> executeToEither(fn: suspend (U) -> Either<R>): JobState<R> = processEither(fn)
```
Comment (williamhogman) [inline src/main/kotlin/JobState.kt]: The operation for some monad M<T> taking function F: T -> M<U>  returning M<U> is usually called a `flatMap`. We haven't followed this convention in other cases so maybe it is not appropriate to use it here. `executeTo$NEW_NAME_OF_EITHER` is fine

File: src/main/kotlin/JobState.kt [hunk 5]
```diff
@@ -189,6 +205,34 @@ class JobState<U: Any> constructor(val value: U?, val interceptors: List<WorkerI
         return process(lastInterceptor, JobStatus.TransientFailure)
     }
 
+    private suspend fun <R: Any> processEither(fn: suspend(U) -> Either<R>, msg: String? = null): JobState<R> {
+        var eitherValue: Either<R>? = null
+        val lastInterceptor = WorkerInterceptor {_, _ ->
+            if(inProgress()) {
+                try{
```
Comment (williamhogman) [inline src/main/kotlin/JobState.kt]: For clarity, avoiding mutable vars, and for avoiding catching exceptions caused by Franz's code I think the approach is more appropriate. 
```
val eitherValue = try {
  fn(this.value!!)
} catch (e: Throwable) {
// Catch stuff here. [...]
}
// Logic handling eitherValue here.. [...]

**Review by williamhogman — CHANGES_REQUESTED** [review]
I think this makes a natural extension to Franz but the naming needs adjusting prior to merging.

**Review by mantono — COMMENTED** [review]
I agree with most of what @williamhogman has written. I would have preferred a solution were WorkerResult could have taken a value instead. Something along
```kotlin
sealed class WorkerResult<T> {
    data class Success<T>(val value: T): WorkerResult()
    object Failure: WorkerResult()
    object Retry: WorkerResult()
}
```

It seems to me that it solves the same issue, right?

**Comment by williamhogman** [comment]
Agree with @mantono

File: src/main/kotlin/WorkerResult.kt [hunk 6]
```diff
@@ -1,7 +1,19 @@
...
+            is WorkerResult.Success<T> -> JobStatus.Incomplete
+            is WorkerResult.Retry -> JobStatus.TransientFailure
+            is WorkerResult.Failure -> JobStatus.PermanentFailure
+        }
+
+    fun value(): T? = when(this){
+        is WorkerResult.Success<T> -> this.value
+        else -> null
+    }
 }
```
Comment (mantono) [inline src/main/kotlin/WorkerResult.kt]: Nice! I don't know if an optional error message (`String?`) or exception (`Throwable?`) would be good to have for `WorkerResult.Retry` and `WorkerResult.Failure` or if it just adds unnecessary complexity?

File: src/main/kotlin/WorkerResult.kt [hunk 7]
```diff
@@ -1,6 +1,23 @@
...
+    val status: WorkerStatus
+) {
+    companion object {
+        fun <L> success(result: L) = WorkerResult(result, WorkerStatus.Success)
+        val retry = WorkerResult(null, WorkerStatus.Retry)
+        val failure = WorkerResult(null, WorkerStatus.Failure)
+    }
+
+    fun toJobStatus(): JobStatus = when(status){
+        WorkerStatus.Success -> JobStatus.Incomplete
```
Comment (mantono) [inline src/main/kotlin/WorkerResult.kt:14]: Is this correct?

File: src/main/kotlin/WorkerResult.kt [hunk 8]
```diff
@@ -1,6 +1,23 @@
 package franz
 
-enum class WorkerResult {
+class WorkerResult<out L> (
+    val value: L? = null,
```
Comment (mantono) [inline src/main/kotlin/WorkerResult.kt:4]: Sealed class was not possible?

File: src/main/kotlin/WorkerResult.kt (hunk 7 above)
Comment (bonahona) [reply src/main/kotlin/WorkerResult.kt:14]: Yes. Incomplete means the pipe continues. A jobstate with the status Success will be considered succesfull and therefor not executing more steps.

**Pull request merged**